BUG: Fixes rounding error in Timestamp.floor()

cbertinato opened the pull request:
- [x] closes #19206 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

jreback commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -462,6 +461,7 @@ Numeric
 - Bug in :class:`DatetimeIndex`, :class:`TimedeltaIndex` addition and subtraction of zero-dimensional integer arrays gave incorrect results (:issue:`19012`)
 - Bug in :func:`Series.__add__` adding Series with dtype ``timedelta64[ns]`` to a timezone-aware ``DatetimeIndex`` incorrectly dropped timezone information (:issue:`13905`)
 - Bug in :func:`Timedelta.__floordiv__` and :func:`Timedelta.__rfloordiv__` dividing by many incompatible numpy objects was incorrectly allowed (:issue:`18846`)
+- Bug in :func:`Timestamp.floor` where time stamps far in the future and past were not rounded correctly (:issue:`19206`)
```
see below, need to add for ceil and for DatetimeIndex floor/ceil as well.

jreback replied:
need tests for DTI as well (these are in pandas/tests/indexes/datetimes/test_ops pls parametrize

jreback replied:
I suspect you need a separate fix there as well.

jreback commented on pandas/_libs/tslibs/timestamps.pyx:
```diff
@@ -600,17 +600,30 @@ class Timestamp(_Timestamp):
+
         if unit < 1000 and unit % 1000 != 0:
             # for nano rounding, work with the last 6 digits separately
             # due to float precision
```
can this case be incorporated as well in the same divisor check?

jreback replied:
maybe set both divisor & buff in a single if that handles really small, really large and other units (divisor=1, buff=1)

jreback replied:
can you do this?

cbertinato replied:
I don't see a simple way of doing this. We would essentially have to move the logic until the arithmetic, which I think would make the code more opaque.  I can move the divisor check into the case were unit >= 1000 and refactor some of it to make it a bit cleaner.

jreback replied:
that's ok, just trying to make it more clear.

jreback commented on pandas/tests/scalar/test_timestamp.py:
```diff
@@ -695,6 +695,27 @@ def test_round(self):
         expected = Timestamp('20130101')
         assert result == expected
 
+        # GH 19206
```
can you add a comment here

jreback commented on pandas/tests/scalar/test_timestamp.py:
```diff
@@ -695,6 +695,27 @@ def test_round(self):
+        expected = Timestamp('1823-01-01 00:00:01.000000010')
+        assert result == expected
+
         # ceil
```
can we test for ceil as well. If you can parametrize this all the better (if not ok too).

Also ok with splitting this giant test to test_round, test_floor, test_ceil

cbertinato replied:
Still working on other change requests, but just wanted to get feedback on refactoring for timestamp tests. There are so many tests that it seemed like a mixed approach of splitting tests and so that failed assertions could be easily traced, and parametrizing to consolidate common code was the way to go.

jreback replied:
see below. we *need* to parameterize as much as possible to cover the cases. and you can still run them individually as well (optionally you can pass ``ids`` to create them)

jreback commented on pandas/tests/scalar/test_timestamp.py:
```diff
@@ -695,6 +695,27 @@ def test_round(self):
+        assert result == expected
+
+        dt = Timestamp('1823-01-01 00:00:01')
+        result = dt.floor('1s')
```
pls test for datetimeindex as well

jreback commented on pandas/tests/scalar/test_timestamp.py:
```diff
@@ -656,61 +657,102 @@ def test_to_pydatetime_nonzero_nano(self):
-        result = dt.round('D')
-        expected = Timestamp('20130202')
-        assert result == expected
+    TStestcase = namedtuple('Tstestcase',
```
pls use parametrize rather rolling your own:http://pandas.pydata.org/pandas-docs/stable/contributing.html#using-pytest

cbertinato commented:
Probably a separate issue, but there appears to be no warning if you specify a Timestamp with a string that causes the underlying int64 to overflow. This works:
```
In [135]: pd.Timestamp(9223372036854775807)
Out[135]: Timestamp('2262-04-11 23:47:16.854775807')

In [136]: pd.Timestamp(9223372036854775808)
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
<ipython-input-136-49b799c43d55> in <module>()
----> 1 pd.Timestamp(9223372036854775808)
pandas/_libs/tslib.pyx in pandas._libs.tslib.Timestamp.__new__ (pandas/_libs/tslib.c:10051)()
pandas/_libs/tslib.pyx in pandas._libs.tslib.convert_to_tsobject (pandas/_libs/tslib.c:27665)()
OverflowError: Python int too large to convert to C long
```
but this is silent:
```
In [138]: pd.Timestamp('2262-04-11 23:47:16.854775807')
Out[138]: Timestamp('2262-04-11 23:47:16.854775807')

In [139]: pd.Timestamp('2262-04-11 23:47:16.854775808')
Out[139]: Timestamp('2262-04-11 23:47:16.854775')
```
Shouldn't this throw an exception?

pep8speaks commented:
Hello @cbertinato! Thanks for updating the PR.



 - In the file [`pandas/tests/indexes/datetimes/test_ops.py`](https://github.com/pandas-dev/pandas/blob/36ec478deaa5e2d6eab3dce1088441f27787d084/pandas/tests/indexes/datetimes/test_ops.py), following are the PEP8 issues :

> [Line 178:1](https://github.com/pandas-dev/pandas/blob/36ec478deaa5e2d6eab3dce1088441f27787d084/pandas/tests/indexes/datetimes/test_ops.py#L178): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

jreback commented:
@cbertinato this is all correct
```
In [1]: pd.Timestamp.max
Out[1]: Timestamp('2262-04-11 23:47:16.854775807')

In [2]: pd.Timestamp.max.value
Out[2]: 9223372036854775807

In [3]: pd.Timestamp(pd.Timestamp.max.value)
Out[3]: Timestamp('2262-04-11 23:47:16.854775807')

In [4]: pd.Timestamp(pd.Timestamp.max.value + 1)
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
<ipython-input-4-cef6abffd699> in <module>()
----> 1 pd.Timestamp(pd.Timestamp.max.value + 1)

~/pandas/pandas/_libs/tslibs/timestamps.pyx in pandas._libs.tslibs.timestamps.Timestamp.__new__()
    579             tz, tzinfo = tzinfo, None
    580 
--> 581         ts = convert_to_tsobject(ts_input, tz, unit, 0, 0)
    582 
    583         if ts.value == NPY_NAT:

~/pandas/pandas/_libs/tslibs/conversion.pyx in pandas._libs.tslibs.conversion.convert_to_tsobject()
    287         else:
    288             ts = ts * cast_from_unit(None, unit)
--> 289             obj.value = ts
    290             dt64_to_dtstruct(ts, &obj.dts)
    291     elif is_float_object(ts):

OverflowError: Python int too large to convert to C long

In [5]: pd.Timestamp('2262-04-11 23:47:16.854775808')
Out[5]: Timestamp('2262-04-11 23:47:16.854775')

In [6]: pd.Timestamp('2262-04-11 23:47:16.854775808').value
Out[6]: 9223372036854775000
```

[138] & [139] might be some nanosecond rounding; I suppose you could make an issue about it

jreback commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -95,18 +95,29 @@ def _round(self, freq, rounder):
+            # GH19206
+            # to deal with round-off when unit is large
+            if unit >= 1e9:
+                divisor = 10 ** int(np.log10(unit / 1e7))
```
we should probably pull out all of this rounding logic to a separate function that can be called on a scalar & and array, so we can use it both for Timestamp and DTI (could do it in this PR or as a followup)

cbertinato replied:
I think that's a good idea. We can do it here while this is still open.

jreback replied:
great!

jreback commented on pandas/tests/indexes/datetimes/test_ops.py:
```diff
@@ -144,6 +144,38 @@ def test_numpy_minmax(self):
+         '1823-01-01 00:00:01.000000020')
+    ])
+    def test_round_ops(self, test_input, rounder, freq, expected):
+        for tz in self.tz:
```
I am not sure why tz was not made into a fixture, but I guess ok here

cbertinato replied:
I'll take care of that too. I don't know why either. Must have overlooked it.

jreback replied:
thanks!

jreback commented on pandas/tests/scalar/test_timestamp.py:
```diff
@@ -734,90 +734,95 @@ def test_pprint(self):
 
+    @pytest.mark.parametrize('test_input, rounder, freq, expected, kw', [
+        ('20130101 09:10:11', 'floor', 'D', '20130101', {}),
+        # GH 19206 - times far in the future and past rounding incorrectly
```
you could just pass None for kw (and then turn it in it into an empty dict in the test)

cbertinato committed:
BUG: Rounding error in `Timestamp.floor` and `DatetimeIndex.floor` for dates far in the future and past (GH19206)

cbertinato commented:
Moved the function common to both Timestamps and DTIs out of the Timestamp round function, but left it in timestamp.pyx. Not sure whether to move it out of there completely. Also moved all tests into the refactored test modules.

jreback commented on pandas/tests/indexes/datetimes/test_scalar_compat.py:
```diff
@@ -126,6 +126,22 @@ def test_round(self, tz):
+         '1823-01-01 00:00:01.000000020'),
+        ('1823-01-01 00:00:01', 'floor', '1s', '1823-01-01 00:00:01'),
+        ('1823-01-01 00:00:01', 'ceil', '1s', '1823-01-01 00:00:01')])
+    def test_ceil_floor_edge(self, tz, test_input, rounder, freq, expected):
```
can you add a NaT and another element here (the 2nd can be the same as the first)

jreback commented on pandas/tests/scalar/timestamp/test_unary_ops.py:
```diff
@@ -93,6 +93,22 @@ def test_round_frequencies(self, freq, expected):
         result = stamp.round(freq=freq)
         assert result == expected
 
+    @pytest.mark.parametrize('test_input, rounder, freq, expected', [
```
can you add a NaT here as well

cbertinato committed:
Added NaT cases to Timestamp and DTI tests

jreback commented:
thanks @cbertinato nice patch! keep em coming!
